Add profiling plugin

danielmitterdorfer opened the pull request:
With this commit we add a new X-Pack plugin that retrieves data for Universal Profiler. This functionality is currently implemented in [Kibana](https://github.com/elastic/kibana/tree/main/x-pack/plugins/profiling) in the function `getExecutablesAndStackTraces` which we replicate here for efficiency purposes. The plugin is off by default and can be enabled by setting `xpack.profiling.enabled` to `true`.

danielmitterdorfer committed:
Add profiling plugin

With this commit we add a new plugin that retrieves data for Universal
Profiler. This functionality is currently implemented in Kibana in the
function `getExecutablesAndStackTraces` which we replicate here for
efficiency purposes.

danielmitterdorfer committed:
Profiling API is a non-operator API

danielmitterdorfer committed:
More tests

danielmitterdorfer committed:
Rename profiler package

danielmitterdorfer committed:
Add IT

danielmitterdorfer committed:
Formatting fixes

danielmitterdorfer committed:
More tests

danielmitterdorfer committed:
Add flag xpack.profiling.enabled

danielmitterdorfer committed:
Don't disable transport action

With this commit we keep the transport action enabled. This is to ensure
operator privileges can be checked in all cases (see
OperatorPrivilegesIT.testEveryActionIsEitherOperatorOnlyOrNonOperator).

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

danielmitterdorfer committed:
Update docs/changelog/91640.yaml

elasticsearchmachine commented:
Hi @danielmitterdorfer, I've created a changelog YAML for you.

danielmitterdorfer committed:
Merge remote-tracking branch 'origin/main' into profiler-plugin

danielmitterdorfer commented:
The test failure on `elasticsearch-ci/part-1` is unrelated to this PR and also reproduces on `main`. I've raised and https://github.com/elastic/elasticsearch/issues/91837 and pushed https://github.com/elastic/elasticsearch/pull/91838 to mute the test.

danielmitterdorfer committed:
Merge remote-tracking branch 'origin/main' into profiler-plugin

javanna commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java:
```diff
@@ -0,0 +1,343 @@
+            this.nodeClient.prepareSearch(mediumDownsampled.getName())
+                .setSize(0)
+                .setQuery(request.getQuery())
+                .setTrackTotalHits(true)
```
I think it would make sense to set the parent task id to the inner search request.

danielmitterdorfer replied:
Done by using `ParentTaskAssigningClient` as you suggested below.

javanna commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/utils/MapExtractor.java:
```diff
@@ -0,0 +1,34 @@
+    private MapExtractor() {}
+
+    @SuppressWarnings("unchecked")
+    public static <T> T read(Map<String, Object> source, String... path) {
```
I am assuming it is not possible to reuse ObjectPath?

danielmitterdorfer replied:
Thanks! I was not aware of that class and replaced all usages of `MapExtractor` now with `ObjectPath`.

javanna commented on x-pack/plugin/profiler/src/test/java/org/elasticsearch/xpack/profiler/GetProfilingRequestTests.java:
```diff
@@ -0,0 +1,102 @@
+
+import static java.util.Collections.emptyList;
+
+public class GetProfilingRequestTests extends ESTestCase {
```
You could probably have extended `AbstractXContentSerializingTestCase` ?

danielmitterdorfer replied:
`GetProfilingRequestTests`  tests different scenarios, e.g. how the parser behaves when a parse error occurs and I'm not certain how that would be handled when inheriting from `AbstractXContentSerializingTestCase`. My understanding is that this only supports a single scenario and I'd probably have to implement one test class per scenario? 

I've also implemented XContent handling "asymmetrically", in the sense that `GetProfilingRequest` only supports parsing (but not rendering an object representation to XContent) and `GetProfilingResponse` only supports rendering but not parsing. Is it common to implement both directions although only one is needed?

javanna replied:
oh I see what you mean, I think that's to do with the high-level client needs, which are now no longer a concern. It may be good to have bi-direction only for testing purposes, but it's overkill I guess.

javanna replied:
One thing I was thinking is that you could also reuse only the Wire part then, and extend AbstractWireSerliazingTestCase if it helps saving boiler plate and possibly extend the coverage too, but up to you, not a requirement.

danielmitterdorfer replied:
Thanks for the suggestion! `GetProfilingResponseTests` was a rather good candidate to inherit from `AbstractWireSerializingTestCase` so I've implemented this in https://github.com/elastic/elasticsearch/pull/91640/commits/84f1acfe8a2133dc833b56d2c7a16d19a227907d.

javanna commented on x-pack/plugin/profiler/src/test/java/org/elasticsearch/xpack/profiler/GetProfilingResponseTests.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.Map;
+import java.util.function.Supplier;
+
+public class GetProfilingResponseTests extends ESTestCase {
```
here too, you could have extended AbstractXContentSerializingTestCase

javanna commented on x-pack/plugin/profiler/src/test/java/org/elasticsearch/xpack/profiler/StackFrameTests.java:
```diff
@@ -0,0 +1,61 @@
+        assertEquals(Integer.valueOf(31733), frame.functionOffset);
+        assertEquals(Integer.valueOf(22), frame.lineNumber);
+    }
+
```
nit: you could test equals and hashcode using EqualsHashCodeTestUtils

danielmitterdorfer replied:
Addressed in https://github.com/elastic/elasticsearch/pull/91640/commits/c21262b78a3faa152e865a7e834ff858c4cc3f10.

javanna commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/StackFrame.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Map;
+import java.util.Objects;
+
+final class StackFrame implements ToXContent {
```
this always has a top level anonymous object, then it should be a ToXContentObject.

danielmitterdorfer replied:
Addressed in https://github.com/elastic/elasticsearch/pull/91640/commits/c21262b78a3faa152e865a7e834ff858c4cc3f10.

javanna commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/StackTrace.java:
```diff
@@ -0,0 +1,231 @@
+import java.util.Arrays;
+import java.util.Map;
+
+final class StackTrace implements ToXContent {
```
ToXContentObject

danielmitterdorfer replied:
Addressed in https://github.com/elastic/elasticsearch/pull/91640/commits/c21262b78a3faa152e865a7e834ff858c4cc3f10.

javanna commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java:
```diff
@@ -0,0 +1,343 @@
+                    .subAggregation(new SumAggregationBuilder("count").field("Stacktrace.count"))
+            )
+            .addAggregation(new SumAggregationBuilder("total_count").field("Stacktrace.count"))
+            .setPreFilterShardSize(1)
```
What is the rationale for always enabling the can match phase?

danielmitterdorfer replied:
Good catch! I took this verbatim from the Kibana plugin (see https://github.com/elastic/kibana/blame/main/x-pack/plugins/profiling/server/routes/stacktrace.ts#L138). I followed up with the team to understand its history and learned that this was introduced when we had used `composite` aggs and it (accidentally?) sped up query execution. So I believe this is obsolete here and will remove this for now.

javanna commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java:
```diff
@@ -0,0 +1,343 @@
+        this.nodeClient.prepareSearch(eventsIndex.getName())
+            .setTrackTotalHits(false)
+            .setQuery(request.getQuery())
+            .addAggregation(
```
I think setting the parent task is should be done in all of the subsearch requests. I am thinking that it's not necessary to have a tree structure, but rather have all of the subsearches be the parent of the main profiling action? In that case you could use a ParentTaskAssigningClient?

danielmitterdorfer replied:
Yes, I think that should be sufficient. I'll use `ParentTaskAssigningClient` then.

javanna commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java:
```diff
@@ -0,0 +1,343 @@
+    protected void doExecute(Task submitTask, GetProfilingRequest request, ActionListener<GetProfilingResponse> submitListener) {
+        try (var ignored = threadContext.newTraceContext()) {
+            EventsIndex mediumDownsampled = EventsIndex.MEDIUM_DOWNSAMPLED;
+            this.nodeClient.prepareSearch(mediumDownsampled.getName())
```
Is this index always local or could it be a remote index? In general, I am assuming that this API will not support cross-cluster execution, but maybe it's a good question to ask.

danielmitterdorfer replied:
Indeed, this index is always local at the moment and there are at least no immediately plans to support cross-cluster search. https://github.com/elastic/elasticsearch/issues/90129 would be required to issue the subsequent `mget` calls in a remote cluster. 

If I understand the reference docs correctly, we'd need to combine the names of the remote clusters and the index name(s) into an index pattern in order to run this search request against multiple clusters?

javanna replied:
yes a search with index `remote:index` goes to the remote cluster configured with `remote` alias.

danielmitterdorfer replied:
Thanks for the confirmation!

javanna reviewed (commented):
I left a couple of comments and questions, nothing major. This looks pretty solid and well tested to me, thanks @danielmitterdorfer .

I would ask @nik9000 to double check the aggs bits in TransportProfilingAction and then we are good?

danielmitterdorfer committed:
Address review feedback

danielmitterdorfer commented:
Thanks for the review @javanna! I addressed your feedback in https://github.com/elastic/elasticsearch/pull/91640/commits/c21262b78a3faa152e865a7e834ff858c4cc3f10. I'm uncertain how to use `AbstractXContentSerializingTestCase` in my setup (see my [comment](https://github.com/elastic/elasticsearch/pull/91640#discussion_r1033292172) above). Also, I clarified that we currently intend that this plugin does not leverage CCS but I'm uncertain what is required to support it (see my other [comment](https://github.com/elastic/elasticsearch/pull/91640#discussion_r1033271785)). Could you be able to take another look?

javanna reviewed (approved):
I left a couple more questions/suggestions, LGTM otherwise

danielmitterdorfer committed:
Implement GetProfilingResponseTests as AbstractWireSerializingTestCase

danielmitterdorfer committed:
Remove obsolete trace context

danielmitterdorfer committed:
Use applyContentParser for request body parsing

danielmitterdorfer commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/GetProfilingRequest.java:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    public void parseXContent(XContentParser parser) throws IOException {
+        if (parser == null) {
```
Yes, indeed. This is because I've used `RestRequest#withContentOrSourceParamParserOrNull()` before. I've now removed this null check and changed to `RestRequest#applyContentParser()` in https://github.com/elastic/elasticsearch/pull/91640/commits/62b9fc891316972d7186c7af8978a7be0fe0ef39 which does not require this check.

danielmitterdorfer commented on x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/RestGetProfilingAction.java:
```diff
@@ -0,0 +1,43 @@
+
+        return channel -> {
+            RestStatusToXContentListener<GetProfilingResponse> listener = new RestStatusToXContentListener<>(channel);
+            RestCancellableNodeClient cancelClient = new RestCancellableNodeClient(client, request.getHttpChannel());
```
Thanks for the pointer! I'll look into that.

danielmitterdorfer replied:
I've added a test now in https://github.com/elastic/elasticsearch/pull/91640/commits/b445b6d67dffe83fca891263e561a0ca1583d505.

danielmitterdorfer committed:
Merge remote-tracking branch 'origin/main' into profiler-plugin

danielmitterdorfer committed:
Add cancellation test

javanna commented on x-pack/plugin/profiler/src/internalClusterTest/java/org/elasticsearch/xpack/profiler/GetProfilingActionIT.java:
```diff
@@ -104,4 +151,137 @@ public void testGetProfilingDataUnfiltered() throws Exception {
+            TaskManager taskManager = internalCluster().getInstance(TransportService.class, nodeName).getTaskManager();
+            Task task = taskManager.getTask(taskId.getId());
+            assertThat(task, instanceOf(CancellableTask.class));
+            assertTrue(((CancellableTask) task).isCancelled());
```
Should we also check that all children tasks are cancelled? My understanding is that we are now checking that the top-level profiling action is cancelled. Ideally, we'd make sure that the children search tasks either don't start before the cancellation, or get cancelled, or complete before the cancellation.

We could check here that there are no task with this specific parent task id, on the other hand that would be the case even if we did not set the parent task id properly. Should we instead try and collect those children tasks before the cancellation and double check that they do get cancelled afterwards?

javanna replied:
I am not entirely sure though where things get blocked with the script below, maybe always at the first inner search request? ideally it would block at any stage but that sounds like it would complicate this test quite a bit?

javanna replied:
looks great!

danielmitterdorfer replied:
Thanks! I'll merge it then when CI is green. :)

danielmitterdorfer committed:
Verify task hierarchy is cancelled
